File: go/models/transport.go
package models

import (
	"challenge-flutter-go/api/utils/geo"
	"errors"
	"time"

	"gorm.io/gorm"
)

type TransportType string

const (
	TransportTypeCar   TransportType = "car"
	TransportTypePlane TransportType = "plane"
	TransportTypeBus   TransportType = "bus"
)

type Transport struct {
	gorm.Model
	TransportType    TransportType `gorm:"not null; default:'car'"`
	StartDate        time.Time     `gorm:"not null"`
	TripID           uint
	Trip             Trip `gorm:"constraint:OnUpdate:CASCADE;not null"`
	AuthorID         uint
	Author           User      `gorm:"constraint:OnUpdate:CASCADE;not null"`
	EndDate          time.Time `gorm:"not null"`
	StartAddress     string    `gorm:"not null"`
	StartLatitude    float64   `gorm:"default:0"`
	StartLongitude   float64   `gorm:"default:0"`
	EndAddress       string    `gorm:"not null"`
	EndLatitude      float64   `gorm:"default:0"`
	EndLongitude     float64   `gorm:"default:0"`
	MeetingAddress   string
	MeetingLatitude  float64 `gorm:"default:0"`
	MeetingLongitude float64 `gorm:"default:0"`
	MeetingTime      time.Time
	Price            float64 `gorm:"not null; default:0" validate:"min=0"`
}

// Checks if the transport type is valid, returns false if it's not & aborts the request
func (transport *Transport) isTransportTypeValid() (isValid bool) {
	return transport.TransportType == TransportTypeCar || transport.TransportType == TransportTypePlane || transport.TransportType == TransportTypeBus
}

func (transport *Transport) BeforeSave(tx *gorm.DB) (err error) {
	if !transport.isTransportTypeValid() {
		return gorm.ErrInvalidData
	}
	if err = geo.GetGeoLocation(transport.StartAddress, &transport.StartLatitude, &transport.StartLongitude); err != nil {
		return errors.New("geo location api is on an error state")
	}
	if err = geo.GetGeoLocation(transport.EndAddress, &transport.EndLatitude, &transport.EndLongitude); err != nil {
		return errors.New("geo location api is on an error state")
	}
	if transport.MeetingAddress != "" {
		if err = geo.GetGeoLocation(transport.MeetingAddress, &transport.MeetingLatitude, &transport.MeetingLongitude); err != nil {
			return errors.New("geo location api is on an error state")
		}
	}
	return
}

File: move_together_app/lib/core/models/transport.dart
import 'package:move_together_app/core/models/user.dart';
import 'package:move_together_app/utils/dynamic_to_double.dart';

enum TransportType {
  car,
  plane,
  bus,
}

class Transport {
  final int id;
  final TransportType transportType;
  final DateTime startDate;
  final DateTime endDate;
  final String startAddress;
  final String endAddress;
  final String? meetingAddress;
  final DateTime? meetingTime;
  final User author;
  final double price;

  Transport({
    required this.id,
    required this.transportType,
    required this.startDate,
    required this.endDate,
    required this.startAddress,
    required this.endAddress,
    this.meetingAddress,
    this.meetingTime,
    required this.author,
    required this.price,
  });

  factory Transport.fromJson(Map<String, dynamic> json) {
    return Transport(
      id: json['id'],
      transportType: TransportType.values.firstWhere((e) => e.toString().split('.').last == json['transportType']),
      startDate: DateTime.parse(json['startDate']),
      endDate: DateTime.parse(json['endDate']),
      startAddress: json['startAddress'],
      endAddress: json['endAddress'],
      meetingAddress: json['meetingAddress'],
      meetingTime: DateTime.parse(json['meetingTime']),
      author: User.fromJson(json['author']),
      price: dynamicToDouble(json['price']),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'transportType': transportType.toString(),
      'startDate': startDate,
      'endDate': endDate,
      'startAddress': startAddress,
      'endAddress': endAddress,
      'meetingAddress': meetingAddress,
      'meetingTime': meetingTime,
      'author': author.toJson(),
      'price': price,
    };
  }
}